File: dotnet/source/cmf.eventing.berico/EventRegistration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Common.Logging;

using cmf.bus;
using cmf.bus.berico;

namespace cmf.eventing.berico
{
    public class EventRegistration : IRegistration
    {
        protected IEventHandler _handler;
        protected IDictionary<string, string> _registrationInfo;
        protected IEnumerable<IInboundEventProcessor> _inboundChain;
        protected ILog _log;


        public virtual Predicate<Envelope> Filter
        {
            get { return null; }
        }

        public virtual IDictionary<string, string> Info
        {
            get { return _registrationInfo; }
        }


        public EventRegistration(IEventHandler handler, IEnumerable<IInboundEventProcessor> inboundChain)
        {
            _handler = handler;
            _inboundChain = inboundChain;

            _registrationInfo = new Dictionary<string, string>();
            _registrationInfo.SetMessageTopic(handler.Topic);

            _log = LogManager.GetLogger(this.GetType());
        }


        public virtual object Handle(Envelope env)
        {
            _log.Debug("Enter Handle");

            object ev = null;
            object result = null;

            if (this.ProcessInbound(ref ev, ref env))
            {
                try
                {

                    result = _handler.Handle(ev, env.Headers);
                }
                catch (Exception ex)
                {
                    _log.Warn("Caught an exception while handling an event", ex);
                    result = this.HandleFailed(env, ex);
                }
            }

            _log.Debug("Leave Handle");
            return result;
        }

        public virtual object HandleFailed(Envelope env, Exception ex)
        {
            _log.Debug("Enter HandleFailed");

            // either log & return or log & throw
            try
            {
                _log.Debug("Leave HandleFailed");
                return _handler.HandleFailed(env, ex);
            }
            catch (Exception failedToFail)
            {
                _log.Error("Caught an exception attempting to raise the Failed event", failedToFail);
                throw;
            }
        }


        protected virtual bool ProcessInbound(ref object ev, ref Envelope env)
        {
            bool processed = true;
            IDictionary<string, object> processorContext = new Dictionary<string, object>();

            try
            {
                foreach (IInboundEventProcessor processor in _inboundChain)
                {
                    if (!processor.ProcessInbound(ref ev, ref env, ref processorContext))
                    {
                        processed = false;
                        _log.Info(string.Format("A processor of type {0} halted the inbound processing chain", processor.GetType().FullName));
                        break;
                    }
                }
            }
            catch (Exception ex)
            {
                _log.Error("Caught an exception while sending an inbound event through the processing chain", ex);
                processed = false;
            }

            return processed;
        }
    }
}

File: java/src/main/java/cmf/eventing/berico/EventRegistration.java
package cmf.eventing.berico;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cmf.bus.Envelope;
import cmf.bus.EnvelopeHeaderConstants;
import cmf.bus.IEnvelopeFilterPredicate;
import cmf.bus.IRegistration;
import cmf.eventing.IEventHandler;

public class EventRegistration implements IRegistration {

    @SuppressWarnings("rawtypes")
	protected IEventHandler eventHandler;
    
    protected IEnvelopeFilterPredicate filterPredicate;
    
    protected List<IInboundEventProcessor> inboundChain;
    
    protected Map<String, String> registrationInfo;

    @SuppressWarnings({ "rawtypes" })
    public EventRegistration(IEventHandler eventHandler, List<IInboundEventProcessor> inboundChain) {

        this.eventHandler = eventHandler;
        this.inboundChain = inboundChain;

        registrationInfo = new HashMap<String, String>();
        registrationInfo.put(EnvelopeHeaderConstants.MESSAGE_TOPIC, eventHandler.getEventType().getCanonicalName());
    }

    @Override
    public IEnvelopeFilterPredicate getFilterPredicate() {
        return filterPredicate;
    }

    @Override
    public Map<String, String> getRegistrationInfo() {
        return registrationInfo;
    }
    
	@Override
	@SuppressWarnings("unchecked")
	public Object handle(Envelope env) throws Exception {
        Object ev = null;
        Object result = null;

        ProcessingContext processorContext = new ProcessingContext(env, ev);
        if (processInbound(processorContext)) {
            try {
                result = eventHandler.handle(processorContext.getEvent(), processorContext.getEnvelope().getHeaders());
            } catch (Exception ex) {
                result = handleFailed(env, ex);
            }
        }

        return result;
    }

    @Override
    public Object handleFailed(Envelope env, Exception ex) throws Exception {
        try {
            return eventHandler.handleFailed(env, ex);
        } catch (Exception failedToFail) {
            throw failedToFail;
        }
    }

    protected boolean processInbound(ProcessingContext processorContext) {
        boolean processed = true;

        try {
            for (IInboundEventProcessor processor : inboundChain) {
                if (!processor.processInbound(processorContext)) {
                    processed = false;
                    break;
                }
            }
        } catch (Exception ex) {
            processed = false;
        }

        return processed;
    }

    public void setFilterPredicate(IEnvelopeFilterPredicate filterPredicate) {
        this.filterPredicate = filterPredicate;
    }

    public void setRegistrationInfo(Map<String, String> registrationInfo) {
        this.registrationInfo = registrationInfo;
    }
}
